plonky2/src/fri/batch_prover.rs

```diff
@@ -0,0 +1,463 @@
+use crate::hash::merkle_tree::MerkleTree;
+use crate::iop::challenger::Challenger;
+use crate::plonk::config::GenericConfig;
+use crate::plonk::plonk_common::reduce_with_powers;
+use crate::timed;
+use crate::util::timing::TimingTree;
+
+/// Builds a batch FRI proof.
+pub fn batch_fri_proof<F: RichField + Extendable<D>, C: GenericConfig<D, F = F>, const D: usize>(
+    initial_merkle_trees: &[&FieldMerkleTree<F, C::Hasher>],
+    lde_polynomial_coeffs: PolynomialCoeffs<F::Extension>,
+    lde_polynomial_values: &mut [PolynomialValues<F::Extension>],
+    challenger: &mut Challenger<F, C::Hasher>,
+    fri_params: &FriParams,
+    timing: &mut TimingTree,
+) -> FriProof<F, C::Hasher, D> {
+    let n = lde_polynomial_coeffs.len();
+    assert_eq!(lde_polynomial_values[0].len(), n);
+    // The polynomial vectors should be sorted by degree, from largest to smallest, with no duplicate degrees.
+    assert!(lde_polynomial_values
+        .windows(2)
+        .all(|pair| { pair[0].len() > pair[1].len() }));
+
+    // Commit phase
+    let (trees, final_coeffs) = timed!(
+        timing,
+        "fold codewords in the commitment phase",
+        batch_fri_committed_trees::<F, C, D>(
+            lde_polynomial_coeffs,
+            lde_polynomial_values,
+            challenger,
+            fri_params,
+        )
+    );
+
+    // PoW phase
+    let pow_witness = timed!(
+        timing,
+        "find proof-of-work witness",
+        fri_proof_of_work::<F, C, D>(challenger, &fri_params.config)
+    );
+
+    // Query phase
+    let query_round_proofs = batch_fri_prover_query_rounds::<F, C, D>(
+        initial_merkle_trees,
+        &trees,
+        challenger,
+        n,
+        fri_params,
+    );
+
+    FriProof {
+        commit_phase_merkle_caps: trees.iter().map(|t| t.cap.clone()).collect(),
+        query_round_proofs,
+        final_poly: final_coeffs,
+        pow_witness,
+    }
+}
+
+pub(crate) fn batch_fri_committed_trees<
+    F: RichField + Extendable<D>,
+    C: GenericConfig<D, F = F>,
+    const D: usize,
+>(
+    mut final_coeffs: PolynomialCoeffs<F::Extension>,
+    values: &[PolynomialValues<F::Extension>],
+    challenger: &mut Challenger<F, C::Hasher>,
+    fri_params: &FriParams,
+) -> FriCommitedTrees<F, C, D> {
+    let mut trees = Vec::with_capacity(fri_params.reduction_arity_bits.len());
+    let mut shift = F::MULTIPLICATIVE_GROUP_GENERATOR;
+    let mut polynomial_index = 1;
+    let mut final_values = values[0].clone();
+    for arity_bits in &fri_params.reduction_arity_bits {
+        let arity = 1 << arity_bits;
+
+        reverse_index_bits_in_place(&mut final_values.values);
+        let chunked_values = final_values
+            .values
+            .par_chunks(arity)
+            .map(|chunk: &[F::Extension]| flatten(chunk))
+            .collect();
+        let tree = MerkleTree::<F, C::Hasher>::new(chunked_values, fri_params.config.cap_height);
+
+        challenger.observe_cap(&tree.cap);
```

Comment:
Can we please get this one ready for parallelising later?

Eg first observe all the trees' caps, then draw a single challenge 'beta' to use for all trees?

Or alternatively (easier to run in parallel, but perhaps less secure), clone the challenger for each run through the loop, and make each tree only observe its own cap. After the loop, we can feed all the caps to the original instance (ie non-cloned version) of the passed in challenger.

Conceptually, I want to decrease the rounds of back-and-forth communication between Arthur and Merlin, by de-interleaving and batching.

Reply:
This is a different 'beta'

Reply:
(I'm checking what we can do.)

Reply:
In order to support batch FRI, I suggest that we postpone the parallelizing changes, since we have multiple joins during the table proving processes.

For example, we need to calculate the Field Merkle Tree for all the table traces immediately after all the table traces are generated.

Reply:
Yes, I don't want to parallelise itself, but I want to see whether we can make the logic ready for parallelising.

However, for this function, after playing with the code for a bit, I'm not sure we can parallelise it easily.  Mostly because the betas are actually dependent on each other.  (Not just accidentally, but in an essential manner.  At least as far as I can tell.)

Reply:
Proving Table 1 -> trace polys \                                   /-> CTL polys \
                                                    (join)->generate FMT                     (join)->generate FMT -> ...
Proving Table 2 -> trace polys /                                  \-> CTL polys /